lib/beacon/beacon.go

```diff
@@ -19,7 +23,84 @@ type recv struct {
 
 type Interface interface {
 	suture.Service
+	fmt.Stringer
 	Send(data []byte)
 	Recv() ([]byte, net.Addr)
 	Error() error
 }
+
+type cast struct {
+	*suture.Supervisor
+	name    string
+	reader  util.ServiceWithError
+	writer  util.ServiceWithError
+	outbox  chan recv
+	inbox   chan []byte
+	stopped chan struct{}
+}
+
+// newCast creates a base object for multi- or broadcasting. Afterwards the
+// caller needs to set reader and writer with the addReader and addWriter
+// methods to get a functional implementation of Interface.
+func newCast(name string) *cast {
+	return &cast{
+		Supervisor: suture.New(name, suture.Spec{
+			// Don't retry too frenetically: an error to open a socket or
+			// whatever is usually something that is either permanent or takes
+			// a while to get solved...
+			FailureThreshold: 2,
+			FailureBackoff:   60 * time.Second,
+			// Only log restarts in debug mode.
+			Log: func(line string) {
+				l.Debugln(line)
+			},
+			PassThroughPanics: true,
+		}),
+		name:    name,
+		inbox:   make(chan []byte),
+		outbox:  make(chan recv, 16),
+		stopped: make(chan struct{}),
+	}
+}
+
+func (c *cast) addReader(r util.ServiceWithError) {
+	c.reader = r
+	c.Add(r)
+}
+
+func (c *cast) addWriter(w util.ServiceWithError) {
+	c.writer = w
+	c.Add(w)
```

Comment:
This could just be:
```
func (c *cast) addWriter(svc func(stop chan struct{}) error) {
    c.writer = util.ServiceWithError(func(stop chan struct) error {
         l.Debug("Starting " + c.name + " writer")
         err := svc(stop)
         l.Debug("Stopped " + c.name + " writer", err)
         return err
    })
    c.Add(c.writer)
}
```

then on both services you just do:
```
        c.addWriter(func(stop chan struct{}) error {
		return readBroadcasts(c.outbox, port, stop)
	})
```


Also, note that reader and writer are the other way around?